brave/src/main/java/brave/propagation/CorrelationFields.java

```diff
@@ -0,0 +1,196 @@
+ *   of spans</li>
+ *   <li>Span-scoped data implies sharing of data must happen after a tracer api. This means metrics
+ *   can't reliably collaborate with this data unless metrics perform after span allocation.</li>
+ *   <li>Tracer apis are more complex than tagging apis, and increases the compatibility complexity
+ *   verses a standalone tagging api.</li>
+ * </ul>
+ * </pre>
+ *
+ * <p>The following are some areas where the Brave can mitigate some of these concerns.
+ * <ul>
+ * <li>Brave routinely attaches trace contexts to a thread via the {@link CurrentTraceContext},
+ * which publishes them visible to other code such as logging or metrics contexts. By attaching
+ * correlation fields to a trace context, as opposed to a span, some visibility races with logging
+ * or metrics are solved.</li>
+ * <li>The overhead problem exists, but it is minimized by setting the default to opt-out of field
+ * propagation. This reduces overhead to the cost of the {@link #NOOP reference}.</li>
+ * <li>The api surface area is limited by introduction of this type, which can be injected where
+ * needed. For example, synchronizing a logging or metrics context needs access to this type</li>
+ * </ul>
+ *
+ * <h3>Application notes</h3>
+ *
+ * <p>Where possible, instrument nearest to the original request. For example, ordering before
+ * metrics can allow any extracted fields to be visible as a metrics dimension or logging context key.
+ *
+ * <p>Do not use correlation fields as ad-hoc storage, as they will add weight to the request and
+ * expose them to arbitrary access. A password would be a very bad example of a correlation field.
+ */
+public abstract class CorrelationFields {
+
+  /** When true, no fields are recorded. Use this flag to avoid performing expensive computation. */
+  public abstract boolean isNoop();
+
+  // cannot depend on java 8 apis
+  public interface Consumer {
+    void accept(String name, String value);
+  }
+
+  public abstract void put(String name, String value);
+
+  /**
+   * Retrieves a correlation key by name or returns null if not found.
+   *
+   * @throws IllegalArgumentException if the input name is not valid
+   */
+  @Nullable public abstract String get(String name);
+
+  /**
+   * Invokes the consumer for each propagation field.
+   *
+   * This is similar to {@link Map#forEach(java.util.function.BiConsumer)} and used to synchronize contexts like MDC.
+   */
+  public abstract void forEach(Consumer consumer);
+
+  public static final CorrelationFields NOOP = new CorrelationFields() {
+
+    @Override public boolean isNoop() {
+      return true;
+    }
+
+    @Override public void put(String name, String value) {
+    }
+
+    @Override public String get(String name) {
+      return null;
+    }
+
+    @Override public void forEach(Consumer consumer) {
+    }
+
+    @Override public String toString() {
+      return "NoopCorrelationFields{}";
+    }
+  };
+
+  static final class Real extends CorrelationFields {
```

Comment:
PS I am starting with LHM eventhough it has more memory cost than array-offset indexing.